### PR TITLE
Support class name/named constructor/class fields contain $ symbols

### PR DESCRIPTION
**Pull request by medz**
- [x]  #833

### CODE REVIEW

**Comment by Woutwo** [comment]
Awesome, just ran in this issue when using generated classes by the graphql codegen

File: packages/freezed/test/integration/concrete.dart [hunk 1]
```diff
@@ -96,3 +96,10 @@ class Const with _$Const {
   const Const._();
   const factory Const(int a) = _Const;
 }
+
+@freezed
+class Dollar with _$Dollar {
```
Comment (rrousselGit) [inline packages/freezed/test/integration/concrete.dart:101]: We'd want to test $ in the middle/end of the variable name too.

And in constructor names

File: packages/freezed/lib/src/templates/prototypes.dart [hunk 2]
```diff
@@ -12,6 +12,32 @@ List<String> parseDecorators(List<ElementAnnotation> metadata) {
   ];
 }
 
+String variable(String name) {
+  return name.contains(r'$') ? '\${$name}' : '\$$name';
+}
+
+String encodeVariable(String name) {
+  final buffer = StringBuffer();
```
Comment (rrousselGit) [inline packages/freezed/lib/src/templates/prototypes.dart:20]: Is there a reason why we can't do `return name.replaceAll(r'$' ,r'\$')`?

File: packages/freezed/test/integration/concrete.dart (hunk 1 above)
Comment (medz) [reply packages/freezed/test/integration/concrete.dart:101]: > We'd want to test $ in the middle/end of the variable name too.
> 
> And in constructor names

@rrousselGit Added, now both named constructors and class names support interspersed $ characters

File: packages/freezed/lib/src/templates/prototypes.dart (hunk 2 above)
Comment (medz) [reply packages/freezed/lib/src/templates/prototypes.dart:20]: Of course there is! If the input string is "a\$$$\$b\$" you? Or accidentally call the wrapper repeatedly! I changed the function name (accurately expressing the intent of the function) in the latest code to be compatible with multiple calls.

**Comment by medz** [comment]
@rrousselGit You need to review the new code.

File: packages/freezed/lib/src/templates/prototypes.dart (hunk 2 above)
Comment (rrousselGit) [reply packages/freezed/lib/src/templates/prototypes.dart:20]: Why would the input contain `\$` in it? I don't think that's a valid character for a property name, right?

**Comment by rrousselGit** [comment]
Some tests seem to be failing. Mind having a look?

**Comment by medz** [comment]
> Some tests seem to be failing. Mind having a look?

My mistake, now all tests are successful

**Comment by andreasgan** [comment]
Anything preventing a merge here @rrousselGit?

File: packages/freezed/test/concrete_test.dart [hunk 3]
```diff
@@ -34,5 +34,12 @@ void main() {
     expect(CustomToString().toString(), '42');
     expect(MixedInToString().toString(), '42');
     expect(BaseToString().toString(), '42');
+    expect(const Dollar($test: 'demo').toString(), r'Dollar($test: demo)');
+    expect(const DollarMiddle(te$st: 'demo').toString(),
+        r'DollarMiddle(te$st: demo)');
+    expect(
+        const DollarEnd(test$: 'demo').toString(), r'DollarEnd(test$: demo)');
```
Comment (rrousselGit) [inline packages/freezed/test/concrete_test.dart:41]:
```suggestion
    expect(
      const DollarMiddle(te$st: 'demo').toString(),
      r'DollarMiddle(te$st: demo)',
    );
    expect(
      const DollarEnd(test$: 'demo').toString(),
      r'DollarEnd(test$: demo)',
    );
```

**Comment by rrousselGit** [comment]
Moving to a separate branch because Github is confused since the original fork has been deleted

**Pull request merged**